virttest/qemu_vm.py

```diff
@@ -1669,8 +1685,22 @@ def sort_key(dev):
                 LOG.warn(txt)
 
         smp_err = ""
+        SMP_PREFER_CORES_VERSION_SCOPE = '[6.2.0, )'
+        #  In the calculation of omitted sockets/cores/threads: we prefer
+        #  sockets over cores over threads before 6.2, while preferring
+        #  cores over sockets over threads since 6.2.
+        vcpu_prefer_sockets = params.get("vcpu_prefer_sockets")
+        if not vcpu_prefer_sockets:
+            if utils_qemu.get_qemu_version(qemu_binary)[0] not in \
+                    VersionInterval(SMP_PREFER_CORES_VERSION_SCOPE):
+                # Prefer sockets over cores before 6.2
+                vcpu_prefer_sockets = "true"
+
+        if vcpu_prefer_sockets:
+            vcpu_prefer_sockets = params.get_boolean(vcpu_prefer_sockets)
```

Comment:
```suggestion
        vcpu_prefer_sockets = params.get_boolean("vcpu_prefer_sockets")
        if not vcpu_prefer_sockets:
            if utils_qemu.get_qemu_version(qemu_binary)[0] not in \
                    VersionInterval(SMP_PREFER_CORES_VERSION_SCOPE):
                # Prefer sockets over cores before 6.2
                vcpu_prefer_sockets = True
```

Reply:
This will be changed by the version information when the user is set to `no`.
The logic here is that the user definition is higher than the version judgment.

Reply:
1. `params.get("vcpu_prefer_sockets")` always return string or `None`, so `bool("no") == True`
2. User control > version control, you can use:
    ```python
    if (utils_qemu.get_qemu_version(qemu_binary)[0] not in 
        VersionInterval(SMP_PREFER_CORES_VERSION_SCOPE)):
        vcpu_prefer_sockets = True
    vcpu_prefer_sockets = params.get_boolean("vcpu_prefer_sockets") and vcpu_prefer_sockets
    ```

Reply:
1. we use `vcpu_prefer_sockets = params.get_boolean("vcpu_prefer_sockets")`
So it can return False
2. Isn't this the same as my current code?

Reply:
```
if not vcpu_prefer_sockets:
if vcpu_prefer_sockets:
```

I wrote this just to clearly show the judgment logic, with and without user configuration. If we want to streamline our code, I can take your second suggestion.

Reply:
In your code, if `vcpu_prefer_sockets` is not defined, then `vcpu_prefer_sockets = True` because `if not vcpu_prefer_sockets:`. But if you defined it, whatever the value is, it always goes to `vcpu_prefer_sockets = params.get_boolean(vcpu_prefer_sockets)`.

Here is some reference, I think you mix the usage of `params.get_boolean`
```python
>>> from virttest import utils_params
>>> params = utils_params.Params({})
>>> params["vcpu_prefer_sockets"] = 'no'
>>> vcpu_prefer_sockets = params.get("vcpu_prefer_sockets")
>>> bool(vcpu_prefer_sockets)
True
>>> params.get_boolean(vcpu_prefer_sockets)
False
>>> params["vcpu_prefer_sockets"] = 'yes'
>>> vcpu_prefer_sockets = params.get("vcpu_prefer_sockets")
>>> bool(vcpu_prefer_sockets)
True
>>> params.get_boolean(vcpu_prefer_sockets)
False
>>> params["vcpu_prefer_sockets"] = 'yes'
>>> params.get_boolean('vcpu_prefer_sockets')
True
>>> params["vcpu_prefer_sockets"] = 'no'
>>> params.get_boolean('vcpu_prefer_sockets')
False
```

Reply:
Good point! corrected, thanks a lot!